runtime/runtime_controller.cc

```diff
@@ -107,8 +138,8 @@ RuntimeController::~RuntimeController() {
   }
 }
 
-bool RuntimeController::IsRootIsolateRunning() const {
-  std::shared_ptr<DartIsolate> root_isolate = root_isolate_.lock();
+bool RuntimeController::IsRootIsolateRunning() {
+  std::shared_ptr<DartIsolate> root_isolate = GetRootIsolate().lock();
```

Comment:
Access to `root_isolate_` directly may escape from creating the isolate, this fragility is a price of this pr. Since it is a private variable, the affection scope is inside `RuntimeController` and it might be acceptable in my personal opinion. I'll add some comments for `root_isolate_` to point it out, is there any better workaround?

If someone calls `GetRootIsolate()`, it means the root isolate is required right now, then it is necessary to create it in a synchronous way if the isolate is not created yet. 
Another reason of moving `GetServiceProtocol()->AddHandler(this, GetServiceProtocolDescription());` is that `RuntimeController` should be accessed only on the UI thread, while originally it was running inside `Shell::Setup()` on the platform thread.

Reply:
> Another reason of moving GetServiceProtocol()->AddHandler(this, GetServiceProtocolDescription()); is that RuntimeController should be accessed only on the UI thread, while originally it was running inside Shell::Setup() on the platform thread.

Good catch, I guess we'll want to fix this regardless if we're going to adopt the bigger solution. We'll have more discussions about whether to accept the price of this pr.